src/software/amazon/ionschema/internal/util/RangeIonTimestampPrecision.kt

```diff
@@ -0,0 +1,93 @@
+package software.amazon.ionschema.internal.util
+
+import software.amazon.ion.IonList
+import software.amazon.ion.IonSymbol
+import software.amazon.ion.IonTimestamp
+import software.amazon.ion.Timestamp
+import software.amazon.ionschema.InvalidSchemaException
+
+/**
+ * Implementation of Range<IonTimestamp> that validates that the precision
+ * of a given timestamp is within the expected range.
+ *
+ * This implementation simply translates the allowed timestamp precisions
+ * to a Range<Int>.
+ *
+ * @see IonTimestampPrecision
+ */
+internal class RangeIonTimestampPrecision (
+        ion: IonList
+) : Range<IonTimestamp> {
+
+    private val delegate: Range<Int>
+
+    init {
+        if (!ion.hasTypeAnnotation("range")) {
+            throw InvalidSchemaException("Invalid timestamp range, missing 'range' annotation:  $ion")
+        }
+        if (ion.size != 2) {
+            throw InvalidSchemaException("Invalid timestamp range, size of list must be 2:  $ion")
+        }
+
+        // convert to an int range
+        // e.g., range::[year, exclusive::millisecond] is translated to range::[-4, exclusive::3]
+        val intRangeIon = ion.system.newEmptyList()
+        intRangeIon.addTypeAnnotation("range")
+        ion.forEachIndexed { index, ionValue ->
+            val isValid = try {
+                    if (ionValue is IonSymbol && !ionValue.isNullValue) {
+                        val itp = IonTimestampPrecision.valueOf(ionValue.stringValue())
+                        val ionInt = when (ionValue.stringValue()) {
+                            "min" -> ion.system.newInt(Int.MIN_VALUE)
+                            "max" -> ion.system.newInt(Int.MAX_VALUE)
+                            else -> ion.system.newInt(itp.id)
+                        }
+                        ionValue.typeAnnotations.forEach { ionInt.addTypeAnnotation(it) }
+                        intRangeIon.add(ionInt)
+                        true
+                    } else {
+                        false
+                    }
+                } catch (e: IllegalArgumentException) {
+                    false
+                }
+
+            if (!isValid) {
+                val end = if (index == 0) {
+                    "lower"
+                } else {
+                    "upper"
+                }
+                throw InvalidSchemaException("Invalid timestamp range $end bound:  $ionValue")
+            }
+        }
+        delegate = RangeFactory.rangeOf<Int>(intRangeIon, RangeType.INT)
+    }
+
+    override fun contains(value: IonTimestamp) = delegate.contains(IonTimestampPrecision.toInt(value))
+}
+
+internal enum class IonTimestampPrecision (val id: Int) {
+    year(-4),
```

Comment:
The reason for the negative id values is mysterious to me and should be removed or the reasons for them documented.

Reply:
These also do not follow the typical uppercase naming convention.

Reply:
Here's an alternative I extracted  from another project of ours and simplified:

```Kotlin
enum class DataType(val typeName: String) {
    BOOLEAN("boolean"), 
    SMALLINT("smallint"),
    INTEGER("integer"),
    FLOAT("float");
    
    companion object {
        private val LOOKUP =
            DataType.values()
                .map { it.typeName to it }
                .toMap()

        fun forTypeName(typeName: String): DataType? = LOOKUP[typeName]
    }
}
```

It's a little more work but prevents the need to violate coding guidelines and has the advantage that the textual representation is now decoupled from the enum value's name.

Reply:
You could also do this if you don't care about decoupling the names:

```Kotlin
    enum class DataType {
        BOOLEAN,
        SMALLINT,
        INTEGER,
        FLOAT;

        companion object {
            private val LOOKUP =
                DataType.values()
                    .map { it.name.toLowerCase() to it }
                    .toMap()

            fun forTypeName(typeName: String): DataType? = LOOKUP[typeName]
        }
    
```

Reply:
The coupling is intentional, and I don't see the value in decoupling in order to adhere to *guidelines*.